packages/flutter_tools/lib/src/version.dart

```diff
@@ -752,24 +752,47 @@ class GitTagVersion {
         _runGit('git fetch $_flutterGit --tags', processUtils, workingDirectory);
       }
     }
-    // `--match` glob must match old version tag `v1.2.3` and new `1.2.3-dev.4.5`
-    return parse(_runGit('git describe --match *.*.* --first-parent --long --tags', processUtils, workingDirectory));
+    final List<String> tags = _runGit(
+      'git tag --contains HEAD', processUtils, workingDirectory).split('\n');
+    // Check first for a stable tag
+    for (final String tag in tags) {
+      final String trimmedTag = tag.trim();
+      if (RegExp(r'^\d+\.\d+\.\d+$').hasMatch(trimmedTag)) {
+        return parse(trimmedTag);
+      }
+    }
+    // Next check for a dev tag
+    for (final String tag in tags) {
+      final String trimmedTag = tag.trim();
+      if (RegExp(r'^\d+\.\d+\.\d+-\d+\.\d+\.pre$').hasMatch(trimmedTag)) {
+        return parse(trimmedTag);
+      }
+    }
+
+    // If we're not currently on a tag, use git describe
+    return parse(
+      _runGit(
+        'git describe --match *.*.*-*.*.pre --first-parent --long --tags',
+        processUtils,
+        workingDirectory,
+      )
+    );
   }
 
-  // TODO(fujino): Deprecate this https://github.com/flutter/flutter/issues/53850
-  /// Check for the release tag format of the form x.y.z-dev.m.n
-  static GitTagVersion parseLegacyVersion(String version) {
+  /// Check for the release tag format of the forms x.y.z or x.y.z-m.n.pre
+  static GitTagVersion parseTaggedVersion(String version) {
     final RegExp versionPattern = RegExp(
-      r'^([0-9]+)\.([0-9]+)\.([0-9]+)(-dev\.[0-9]+\.[0-9]+)?-([0-9]+)-g([a-f0-9]+)$');
-    final List<String> parts = versionPattern.matchAsPrefix(version)?.groups(<int>[1, 2, 3, 4, 5, 6]);
+      r'^(\d+)\.(\d+)\.(\d+)(-\d+\.\d+\.pre)?$');
+    final Match match = versionPattern.matchAsPrefix(version);
+    final List<String> parts = match?.groups(<int>[1, 2, 3, 4]);
     if (parts == null) {
       return const GitTagVersion.unknown();
     }
-    final List<int> parsedParts = parts.take(5).map<int>(
+    final List<int> parsedParts = parts.map<int>(
       (String source) => source == null ? null : int.tryParse(source)).toList();
     List<int> devParts = <int>[null, null];
     if (parts[3] != null) {
```

Comment:
This method is confusing, any way we can name the groups since parts[3] is group 4?  Or create locals like:
```
final String majorVersion = match?.group(1);
final String minorVersion = match?.group(2);
final String patchVersion = match?.group(3);
final String preSuffix = match?.group(4);
```

Reply:
done